audiostream/src/ni/media/audio/source/gstreamer_file_source.cpp

```diff
@@ -0,0 +1,352 @@
+{
+    gst_element_set_state( m_pipeline.get(), GST_STATE_NULL );
+    wait_for_async_operation();
+    m_pipeline.reset();
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+GstState gstreamer_file_source::wait_for_async_operation()
+{
+    GstState current_state = GST_STATE_VOID_PENDING;
+    GstState pending_state = GST_STATE_VOID_PENDING;
+
+    while ( gst_element_get_state( m_pipeline.get(), &current_state, &pending_state, GST_MSECOND )
+            == GST_STATE_CHANGE_ASYNC )
+    {
+        g_main_context_iteration( nullptr, FALSE );
+    }
+
+    return current_state;
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::init_gstreamer()
+{
+    static bool gstreamer_initialized = false;
+
+    if ( !std::exchange( gstreamer_initialized, true ) )
+    {
+        gst_init( 0, nullptr );
+    }
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::setup_source( const std::string& path, audio::ifstream_info::container_type container )
+{
+    m_sink.reset( prepare_pipeline( path ) );
+    auto sinkpad     = gst_element_get_static_pad( m_sink.get(), "sink" );
+    auto caps        = gst_pad_get_current_caps( sinkpad );
+    auto caps_struct = gst_caps_get_structure( caps, 0 );
+    fill_format_info( caps_struct, container );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+GstElement* gstreamer_file_source::prepare_pipeline( const std::string& path )
+{
+    m_pipeline.reset( gst_pipeline_new( "pipeline" ) );
+
+    GstElement* source    = gst_element_factory_make( "filesrc", "source" );
+    GstElement* decodebin = gst_element_factory_make( "decodebin", "decoder" );
+    GstElement* queue     = gst_element_factory_make( "queue", "queue" );
+    GstElement* sink      = gst_element_factory_make( "appsink", "sink" );
+
+    gst_bin_add_many( GST_BIN( m_pipeline.get() ), source, decodebin, queue, gst_object_ref( sink ), nullptr );
+    gst_element_link_many( source, decodebin, NULL );
+    gst_element_link_many( queue, sink, NULL );
+
+    g_object_set( source, "location", path.c_str(), NULL );
+    g_object_set( sink, "sync", FALSE, NULL );
+
+    g_signal_connect( decodebin, "pad-added", G_CALLBACK( &onPadAdded ), queue );
+
+    preroll_pipeline();
+    return sink;
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::preroll_pipeline()
+{
+    gst_element_set_state( m_pipeline.get(), GST_STATE_PAUSED );
+
+    if ( wait_for_async_operation() != GST_STATE_PAUSED )
+        throw std::runtime_error( "gstreamer_file_source: pipeline doesn't preroll into paused state" );
```

Comment:
This line randomly throws. There seems to be a race condition when setting / checking the state